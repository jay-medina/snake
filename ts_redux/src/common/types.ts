import { MapStateToProps as MSTP } from 'react-redux';
import { ThunkDispatch, ThunkAction } from 'redux-thunk';
import { AppAction } from '../store/actions';
import { Store } from 'redux';

export type GameThunkAction = ThunkAction<void, AppState, {}, AppAction>;

export type GameThunkDispatch = ThunkDispatch<AppState, {}, AppAction>;

export type DTP<DispatchProps, OwnProps = {}> = (
  dispatch: GameThunkDispatch,
  ownProps: OwnProps,
) => DispatchProps;

export type STP<StateProps, OwnProps = {}> = MSTP<StateProps, OwnProps, AppState>;

export type GameStore = Store<AppState, AppAction> & {
  dispatch: GameThunkDispatch;
};

export type GameState = 'Start' | 'Run' | 'GameOver';

export interface AppState {
  gameState: GameState;
  apple: Apple;
  snake: Snake;
  dimensions: {
    rows: number;
    cols: number;
  };
  score: Score;
}

export interface GridItem {
  row: number;
  col: number;
}

export type Apple = GridItem;

export type Direction = 'left' | 'right' | 'up' | 'down';

export interface Score {
  current: number;
  high: number;
}

export interface Snake {
  body: GridItem[];
  incrementTimer: number;
  lastTimestamp: number;
  direction: Direction;
  newDirection?: Direction;
}

export type TimeStamp = number;
