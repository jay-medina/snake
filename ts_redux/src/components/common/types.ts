export interface GridItem {
  row: number;
  col: number;
}

export type Apple = GridItem;

export type Snake = GridItem[];

export type GameState = 'start' | 'run' | 'gameover';
