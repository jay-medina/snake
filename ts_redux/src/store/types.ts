import { MapStateToProps as MSTP } from 'react-redux';
import { AppState } from './util';
import { ThunkDispatch, ThunkAction } from 'redux-thunk';
import { AppAction } from './actions';

export type GameThunkAction<ReturnValue> = ThunkAction<ReturnValue, AppState, {}, AppAction>;

export type GameThunkDispatch = ThunkDispatch<AppState, {}, AppAction>;

export type DTP<DispatchProps, OwnProps = {}> = (
  dispatch: GameThunkDispatch,
  ownProps: OwnProps,
) => DispatchProps;

export type STP<StateProps, OwnProps = {}> = MSTP<StateProps, OwnProps, AppState>;
