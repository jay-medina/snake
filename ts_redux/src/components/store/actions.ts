// import { GameThunkAction } from './types';
// import { AppState } from './util';

interface GameAction {
  type: 'START_GAME';
}

export type AppAction = GameAction;

export const startGame = (): GameAction => ({
  type: 'START_GAME',
});

// export function randomizeApple(appState: AppState) {

// }

// export const startGameThunk = (): GameThunkAction => (dispatch) => {
//   // randomize the apple position
//   // start the game
// //  return dispatch(startGame);
// };
