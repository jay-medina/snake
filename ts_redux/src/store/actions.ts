import { GameThunkAction, Apple, TimeStamp, Direction } from '../common/types';
import { findNewApplePosition, isSnakeAtPosition } from '../common/util';

interface GameAction {
  type: 'START_GAME';
}

interface AppleAction {
  type: 'UPDATE_APPLE';
  payload: {
    apple: Apple;
  };
}

interface ScoreAction {
  type: 'INCREASE_SCORE';
}

export interface TickTimeAction {
  type: 'TICK_TIME';
  payload: {
    timestamp: TimeStamp;
  };
}

export interface SnakeDirectionAction {
  type: 'UPDATE_SNAKE_DIRECTION';
  payload: {
    direction: Direction;
  };
}

export interface SnakeSizeAction {
  type: 'UPDATE_SNAKE_SIZE';
}

export type SnakeAction = SnakeDirectionAction | SnakeSizeAction;

export type AppAction = GameAction | AppleAction | TickTimeAction | SnakeAction | ScoreAction;

export const startGame = (): GameAction => ({
  type: 'START_GAME',
});

export const newApplePosition = (position: Apple): AppleAction => ({
  type: 'UPDATE_APPLE',
  payload: {
    apple: position,
  },
});

export const tickForward = (timestamp: TimeStamp): TickTimeAction => ({
  type: 'TICK_TIME',
  payload: {
    timestamp,
  },
});

export const growSnake = (): SnakeSizeAction => ({
  type: 'UPDATE_SNAKE_SIZE',
});

export const updateScore = (): ScoreAction => ({
  type: 'INCREASE_SCORE',
});

export const newAppleThunk = (): GameThunkAction => (dispatch, getState) => {
  const appState = getState();

  const newApple = findNewApplePosition(appState);
  return dispatch(newApplePosition(newApple));
};

export const tickForwardThunk = (timestamp: TimeStamp): GameThunkAction => (dispatch, getState) => {
  dispatch(tickForward(timestamp));

  const appState = getState();
  const { snake, apple } = appState;

  if (isSnakeAtPosition(snake.body)(apple)) {
    dispatch(growSnake());
    dispatch(newAppleThunk());
    dispatch(updateScore());
  }
};

export const startGameThunk = (): GameThunkAction => (dispatch) => {
  dispatch(startGame());
  dispatch(newAppleThunk());
};

export const updateSnakeDirection = (direction: Direction): SnakeDirectionAction => ({
  type: 'UPDATE_SNAKE_DIRECTION',
  payload: {
    direction,
  },
});
