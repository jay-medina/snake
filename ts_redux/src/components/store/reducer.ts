import { AppAction } from './actions';
import { AppState, GameState, initialState } from './util';

type Game = (start: AppState | undefined, action: AppAction) => AppState;

export const game: Game = (state = initialState, action) => {
  if (action.type === 'START_GAME') {
    return {
      gameState: GameState.Run,
    };
  }

  return state;
};
