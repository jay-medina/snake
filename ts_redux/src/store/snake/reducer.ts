import { Snake } from '../../common/types';
import { AppAction } from '../actions';
import { isSnakeAbleToMove, moveSnake, isOppositeDirection } from './util';

function updateDirection(snake: Snake): Snake {
  if (snake.newDirection) {
    return {
      ...snake,
      direction: snake.newDirection,
      newDirection: undefined,
    };
  }

  return snake;
}

// TODO determine if snake is ded
export const snakeReducer = (snake: Snake, action: AppAction): Snake => {
  if (action.type === 'TICK_TIME') {
    const { payload } = action;
    const newSnake = updateDirection(snake);

    if (isSnakeAbleToMove(newSnake, payload.timestamp)) {
      return {
        ...newSnake,
        body: moveSnake(newSnake),
        lastTimestamp: payload.timestamp,
      };
    }
  }

  if (action.type === 'UPDATE_SNAKE_DIRECTION') {
    const { payload } = action;

    if (!isOppositeDirection(snake.direction, payload.direction)) {
      return {
        ...snake,
        newDirection: payload.direction,
      };
    }
  }

  if (action.type === 'UPDATE_SNAKE_SIZE') {
    const newSnake = updateDirection(snake);
    const { body } = newSnake;
    const lastPosition = body[body.length - 1];

    return {
      ...newSnake,
      body: [...body, lastPosition],
    };
  }

  return snake;
};
