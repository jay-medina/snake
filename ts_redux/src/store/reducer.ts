import { AppAction } from './actions';
import { GameState, Apple, AppState, Score } from '../common/types';
import { snakeReducer } from './snake/reducer';

const initialState: AppState = {
  score: {
    current: 0,
    high: 100,
  },
  dimensions: {
    cols: 25,
    rows: 25,
  },
  apple: {
    row: 2,
    col: 2,
  },
  snake: {
    body: [{ row: 5, col: 3 }, { row: 5, col: 2 }, { row: 5, col: 1 }],
    incrementTimer: 200,
    lastTimestamp: 0,
    direction: 'right',
  },
  gameState: 'Start',
};

const gameState = (state: GameState, action: AppAction): GameState => {
  if (action.type === 'START_GAME') {
    return 'Run';
  }

  return state;
};

const apple = (state: Apple, action: AppAction): Apple => {
  if (action.type === 'UPDATE_APPLE') {
    return action.payload.apple;
  }

  return state;
};

const score = (state: Score, action: AppAction): Score => {
  if (action.type === 'INCREASE_SCORE') {
    const { current, high } = state;
    const newCurrent = current + 10;

    return {
      current: newCurrent,
      high: high > newCurrent ? high : newCurrent,
    };
  }

  return state;
};

export const app = (state: AppState = initialState, action: AppAction): AppState => {
    return {
      ...state,
      snake: snakeReducer(state.snake, action),
      apple: apple(state.apple, action),
      gameState: gameState(state.gameState, action),
      score: score(state.score, action),
    };
};
