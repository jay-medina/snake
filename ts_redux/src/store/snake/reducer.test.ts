import { snakeReducer } from './reducer'
import { Snake } from '../../common/types'
import { TickTimeAction, tickForward, updateSnakeDirection } from '../actions'

describe('snake', () => {
  let snakey: Snake

  beforeEach(() => {
    snakey = {
      body: [{ row: 5, col: 3 }, { row: 5, col: 2 }, { row: 5, col: 1 }],
      incrementTimer: 200,
      lastTimestamp: 0,
      direction: 'right',
    }
  })

  it('initializes the snake', () => {
    const action: any = {}

    const snake = snakeReducer(snakey, action)
    expect(snake).toEqual(snakey)
  })

  describe('when the time ticks forward', () => {
    let action: TickTimeAction

    beforeEach(() => {
      action = tickForward(10)
    })

    describe('when the snake is not able to move', () => {
      it('returns back the previous state', () => {
        snakey.lastTimestamp = 10
        const nextSnake = snakeReducer(snakey, action)
        expect(nextSnake).toBe(snakey)
      })
    })

    describe('when the snake is able to move', () => {
      let nextSnake: Snake
      const timestamp = 250

      beforeEach(() => {
        snakey.lastTimestamp = 10
        nextSnake = snakeReducer(snakey, tickForward(timestamp))
      })

      it('updates the snakes body', () => {
        expect(nextSnake.body).toMatchSnapshot()
      })

      it('updates the last timestamp', () => {
        expect(nextSnake.lastTimestamp).toBe(timestamp)
      })
    })
  })

  describe('when the snakes direction is updated', () => {
    let nextSnake: Snake

    describe('when the new direction is direct opposite', () => {
      it('returns back the previous state', () => {
        nextSnake = snakeReducer(snakey, updateSnakeDirection('left'))
        expect(nextSnake).toBe(snakey)
      })
    })

    it('updates the direction when its any other direction', () => {
      nextSnake = snakeReducer(snakey, updateSnakeDirection('up'))
      expect(nextSnake.direction).toBe('up')
    })
  })
})
