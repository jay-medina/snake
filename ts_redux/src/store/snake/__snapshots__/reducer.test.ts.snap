// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snake when the time ticks forward when the snake is able to move updates the snakes body 1`] = `
Array [
  Object {
    "col": 4,
    "row": 5,
  },
  Object {
    "col": 3,
    "row": 5,
  },
  Object {
    "col": 2,
    "row": 5,
  },
]
`;
